File: google/array/best_time_to_buy_and_sell_stock.js
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/564/
function maxProfit(prices) {
    const len = prices.length;
    let profit = 0;
    let buyIndex = -1;
    let i = 0;

    if (len === 1) return profit;

    while (i <= len - 2) {
        const isInUpTrend = prices[i] < prices[i + 1];
        const isInDownTrend = prices[i] > prices[i + 1];

        // price moving up
        if (isInUpTrend && buyIndex === -1) buyIndex = i; // not bought yet

        // price moving down
        if (isInDownTrend && buyIndex !== -1) {
            // bought already
            profit += prices[i] - prices[buyIndex];
            buyIndex = -1;
        }
        i++;
    }

    if (buyIndex !== -1) {
        profit += prices[i] - prices[buyIndex];
    }

    return profit;
}

function main() {
    const ip_arr = [7, 1, 5, 3, 6, 4]; // 7
    // const ip_arr = [7, 6, 4, 3, 1]; // 0
    // const ip_arr = [1, 2, 3, 4, 5]; // 4
    // const ip_arr = [5, 4, 3, 2, 1];
    // const ip_arr = [1, 2, 3, 4, 5, 5, 6]; // 4
    // const ip_arr = [1, 2];
    // const ip_arr = [2, 1, 4];

    console.log(maxProfit(ip_arr));
}

main();

File: masteringDsAlgo/src/google/array/BestTimeToBuyAndSellStock.java
package google.array;

public class BestTimeToBuyAndSellStock {

    private int maxProfit(int[] prices) {
        int len = prices.length;
        int profit = 0;
        int buyIndex = -1;
        int currentIndex = 0;

        if (len == 1) return profit;

        while(currentIndex < len - 1) {
            boolean isInUpTrend = prices[currentIndex] < prices[currentIndex+1];
            boolean isInDownTrend = prices[currentIndex] > prices[currentIndex+1];

            if (isInUpTrend && buyIndex == -1) {
                // not bought yet
                buyIndex = currentIndex;
            }

            if (isInDownTrend && buyIndex != -1) {
                // already bought
                profit += prices[currentIndex] - prices[buyIndex];
                buyIndex = -1;
            }

            currentIndex++;
        }

        if (buyIndex != -1) {
            profit += prices[currentIndex] - prices[buyIndex];
        }

        return profit;
    }

    public static void main(String[] args) {
        BestTimeToBuyAndSellStock bestTimeToBuyAndSellStock = new BestTimeToBuyAndSellStock();
        // int prices[] = {7, 1, 5, 3, 6, 4}; // 7
        // int prices[] = {7, 6, 4, 3, 1}; // 0
        // int prices[] = {1, 2, 3, 4, 5}; // 4
        // int prices[] = {5, 4, 3, 2, 1};
        // int prices[] = {1, 2, 3, 4, 5, 5, 6}; // 5
        // int prices[] = {1, 2};
        int[] prices = {2, 1, 4};
        System.out.println(bestTimeToBuyAndSellStock.maxProfit(prices));
    }
}
